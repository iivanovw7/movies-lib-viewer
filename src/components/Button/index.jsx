/**
 * Module contains button component
 * @module components/Button
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import * as R from 'ramda';
import React from 'react';

import { isUndefined } from '../../utils/helpers';

import StyledButton from './StyledButton';
import StyledLink from './StyledLink';

/**
 * Creates styled button component.
 * @name Components/Button
 * @method
 *
 * @param {Object.<module:components/Button~propTypes>} props
 *  contains component props
 *  @see {@link module:components/Button~propTypes}
 * @return {Node} React component with children.
 * @constructor
 */
const Button = (props) => {
  const { variant, children, styling, onClick, index } = props;

  /* eslint-disable react/jsx-props-no-spreading */
  const ButtonElement = R.ifElse(
    R.pipe(R.prop('onClick'), isUndefined),
    R.always(<StyledLink {...props} />),
    R.always(<StyledButton {...props} />),
  );
  /* eslint-enable react/jsx-props-no-spreading */

  return (
    <ButtonElement data-index={index} variant={variant} styling={styling} onClick={onClick}>
      {children}
    </ButtonElement>
  );
};

/**
 * @name propTypes
 * @type {Object}
 * @param {Object} props - React PropTypes
 * @property {string} [props.variant = 'primary'] - color variant string.
 * @property {string | Image} props.children - button children element or text.
 * @property {Array.<string>} [props.styling = []] - image additional styles.
 * @property {Function} props.onClick - click handler.
 * @property {string} props.href - target link.
 * @property {string} [props.target = '_blank'] - target prop.
 * @property {boolean} [props.hidden = false] - if element should be hidden.
 * @property {number} [props.index = 0] - index which is passed in `data-index` attribute.
 * @return {Array} React propTypes
 */
Button.propTypes = {
  variant: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.node])),
  ]),
  styling: PropTypes.array,
  onClick: PropTypes.func,
  href: PropTypes.string,
  target: PropTypes.string,
  hidden: PropTypes.bool,
  index: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

Button.defaultProps = {
  target: '_blank',
  hidden: false,
  index: 0,
};

export default Button;
