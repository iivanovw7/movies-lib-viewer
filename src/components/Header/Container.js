/**
 * Module contains application header container
 * @module components/Header/Container
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const Container = styled.div`
  align-items: center;
  display: flex;
  flex-direction: row;
  height: 100%;
  justify-content: flex-start;
  width: 100%;
`;

export default Container;
