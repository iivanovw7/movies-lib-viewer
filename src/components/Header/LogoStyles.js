/**
 * Module contains additional styles used in logo image.
 * @module components/TopBar/LogoLogoStyles
 * @author Igor Ivanov
 */
import { css } from 'styled-components';

const LogoStyles = css`
  margin-left: 0.8em;

  svg {
    height: 3rem;
    max-height: 48px;
    max-width: 48px;
    width: 3rem;
  }
`;

export default LogoStyles;
