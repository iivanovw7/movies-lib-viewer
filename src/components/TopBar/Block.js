/**
 * Module contains content block in the right side of the container
 * @module components/TopBar/Block
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const Block = styled.div`
  align-items: center;
  background-color: transparent;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

export default Block;
