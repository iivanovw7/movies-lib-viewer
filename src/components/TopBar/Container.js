/**
 * Module contains application TopBar container
 * @module components/TopBar/Container
 * @author Igor Ivanov
 */
import styled from 'styled-components';

import { base } from '../../config/styles';
import { zIndex } from '../../style/mixins';
import { topBarBg } from '../../style/theme/background';
import { topBarBorderColor } from '../../style/theme/borders';

const Container = styled.div`
  align-items: center;
  background-color: ${topBarBg};
  border-bottom: 1px solid ${topBarBorderColor};
  display: flex;
  flex-direction: row;
  height: ${base.topBarHeight}em;
  justify-content: space-between;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: ${zIndex.TopBar};
`;

export default Container;
