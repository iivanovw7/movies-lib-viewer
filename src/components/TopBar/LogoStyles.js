/**
 * Module contains additional styles used in logo image.
 * @module components/TopBar/LogoLogoStyles
 * @author Igor Ivanov
 */
import { css } from 'styled-components';

import { respondToMedia, mediaKey } from '../../style/mixins';

const LogoStyles = css`
  height: 0.8em;
  margin-left: 0.8em;

  ${respondToMedia[mediaKey('xs')]`
    height: 0.6em;
  `}
`;

export default LogoStyles;
