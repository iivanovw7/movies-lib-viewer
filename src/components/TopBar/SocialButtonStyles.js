/**
 * Module contains additional styles used in logo image.
 * @module components/TopBar/SocialButtonStyles
 * @author Igor Ivanov
 */
import { lighten } from 'polished';
import { css } from 'styled-components';

import { colorSet } from '../../config/styles';
import { styleMixins } from '../../style/mixins';

const SocialButtonStyles = css`
  color: ${colorSet.topBarLinks};
  margin: 0.3em 0.3em 0.3em 0.8em;
  ${styleMixins.focusOutline(colorSet.topBarLinks)};

  &:hover {
    color: ${lighten(0.06, colorSet.topBarLinks)};
  }
`;

export default SocialButtonStyles;
