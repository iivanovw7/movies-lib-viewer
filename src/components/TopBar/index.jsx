/**
 * Module contains application top bar.
 * @module components/TopBar
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import React from 'react';

import { ReactComponent as FacebookLogo } from '../../../assets/svg/social_facebook.svg';
import { ReactComponent as InstagramLogo } from '../../../assets/svg/social_instagram.svg';
import { ReactComponent as TwitterLogo } from '../../../assets/svg/social_twitter.svg';
import settings from '../../config/data';
import Button from '../Button';

import Block from './Block';
import Container from './Container';
import SocialButtonStyles from './SocialButtonStyles';

/**
 * Creates TopBar component.
 * @name Components/TopBar
 * @method
 *
 * @param {Object.<module:components/TopBar~propTypes>} props
 *  contains component props
 *  @see {@link module:components/TopBar~propTypes}
 * @return {Node} React component with children.
 * @constructor
 */
function TopBar(props) {
  const { children } = props;
  const { social } = settings;

  return (
    <Container>
      <Block>
        <Button href={social.facebook} target="_self" styling={SocialButtonStyles}>
          <FacebookLogo fill="currentColor" />
        </Button>
        <Button href={social.instagram} target="_self" styling={SocialButtonStyles}>
          <InstagramLogo fill="currentColor" />
        </Button>
        <Button href={social.twitter} target="_self" styling={SocialButtonStyles}>
          <TwitterLogo fill="currentColor" />
        </Button>
      </Block>
      <Block>{children}</Block>
    </Container>
  );
}

/**
 * @name LocaleProvider.propTypes
 * @type {Object}
 * @param {Object} props - React PropTypes
 * @property {Image} props.children - collection of children components.
 * @return {Array} React propTypes
 */
TopBar.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.node])),
  ]),
};

export default TopBar;
