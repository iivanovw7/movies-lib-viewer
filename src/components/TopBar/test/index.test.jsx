/**
 * Module contains TopBar component test suite
 */
import { mount, shallow } from 'enzyme';
import React from 'react';

import ChildMock from '../../../__mocks__/childMock';
import { testName } from '../../../__utils__/utils';
import Button from '../../Button';
import Img from '../../Img';
import Block from '../Block';
import Container from '../Container';
import TopBar from '../index';
import LogoStyles from '../LogoStyles';

/* eslint-disable require-jsdoc, react/jsx-props-no-spreading */
// prettier-ignore
describe(testName('TopBar', 'component test suite'), function TopBarTestSuite() {
  const Composition = (props, children, shouldMount = false) => {
    const node = (<TopBar {...props}>{children}</TopBar>);

    return shouldMount ? mount(node) : shallow(node);
  };

  it('Should match snapshot', () => {
    const component = Composition({}, ChildMock.component);

    expect(component)
      .toMatchSnapshot();
  });

  it('Should render Container', () => {
    const component = Composition({}, ChildMock.component);

    expect(component.find(Container))
      .toHaveLength(1);
  });

  it('Should render Block', () => {
    const component = Composition({}, ChildMock.component);

    expect(component.find(Block))
      .toHaveLength(1);
  });

  it('Should render child mock', () => {
    const component = Composition({}, ChildMock.component);
    const child = component.find(`#${ChildMock.id}`);

    expect(child)
      .toHaveLength(1);

    expect(child.text())
      .toEqual(ChildMock.child);
  });

  it(testName('Button', 'Should render <Button /> with props'), () => {
    const component = Composition({}, ChildMock.component, true);
    const button = component.find(Button);
    const image = component.find(Img);

    expect(button)
      .toHaveLength(1);

    expect(button.prop('variant'))
      .toEqual('primary');

    expect(button.prop('href'))
      .toEqual('https://www.themoviedb.org/?language=en-US');

    expect(image)
      .toHaveLength(1);

    expect(image.prop('alt'))
      .toEqual('TMDB');

    expect(image.prop('styling'))
      .toBe(LogoStyles);
  });
});
/* eslint-enable require-jsdoc, react/jsx-props-no-spreading */
