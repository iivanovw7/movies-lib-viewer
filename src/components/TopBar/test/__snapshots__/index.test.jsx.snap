// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[33mTopBar[39m]: component test suite Should match snapshot 1`] = `
<Container>
  <Button
    hidden={false}
    href="https://www.themoviedb.org/?language=en-US"
    target="_blank"
    variant="primary"
  >
    <Img
      alt="TMDB"
      className=""
      src="SvgrURL"
      styling={
        Array [
          "height:0.8em;margin-left:0.8em;",
          "@media (min-width:",
          "480",
          "px){",
          "
    height: 0.6em;
  ",
          ";}",
        ]
      }
    />
  </Button>
  <Block>
    <div
      id="child-mock-id"
    >
      child-mock-id
    </div>
  </Block>
</Container>
`;
